src/instance.rs

```diff
@@ -357,47 +118,15 @@ impl<T> Py<T> {
     }
 }
 
-/// Specialization workaround
-trait AsPyRefDispatch<T: PyTypeInfo>: AsPyPointer {
-    fn as_ref_dispatch(&self, _py: Python) -> &T;
-    fn as_mut_dispatch(&mut self, _py: Python) -> &mut T;
-}
-
-impl<T: PyTypeInfo> AsPyRefDispatch<T> for Py<T> {
-    default fn as_ref_dispatch(&self, _py: Python) -> &T {
-        unsafe {
-            let ptr = (self.as_ptr() as *mut u8).offset(T::OFFSET) as *mut T;
-            ptr.as_ref().expect("Py has a null pointer")
-        }
-    }
-    default fn as_mut_dispatch(&mut self, _py: Python) -> &mut T {
-        unsafe {
-            let ptr = (self.as_ptr() as *mut u8).offset(T::OFFSET) as *mut T;
-            ptr.as_mut().expect("Py has a null pointer")
-        }
-    }
-}
-
-impl<T: PyTypeInfo + PyNativeType> AsPyRefDispatch<T> for Py<T> {
-    fn as_ref_dispatch(&self, _py: Python) -> &T {
-        unsafe { &*(self as *const instance::Py<T> as *const T) }
-    }
-    fn as_mut_dispatch(&mut self, _py: Python) -> &mut T {
-        unsafe { &mut *(self as *mut _ as *mut T) }
-    }
+pub trait AsPyRef<T: PyTypeInfo>: Sized {
+    /// Return reference to object.
+    fn as_ref(&self, py: Python) -> &T;
 }
 
-impl<T> AsPyRef<T> for Py<T>
-where
-    T: PyTypeInfo,
-{
-    #[inline]
-    fn as_ref(&self, py: Python) -> PyRef<T> {
-        PyRef::from_ref(self.as_ref_dispatch(py))
-    }
-    #[inline]
-    fn as_mut(&mut self, py: Python) -> PyRefMut<T> {
-        PyRefMut::from_mut(self.as_mut_dispatch(py))
+impl<T: PyTypeInfo> AsPyRef<T> for Py<T> {
+    fn as_ref(&self, _py: Python) -> &T {
```

Comment:
Given that `Python` is unused here, what about a blanket impl

    impl<T: AsRef<T>> AsPyRef<T> for T 

and then implementing `AsRef<T> for Py<T>`.

Reply:
This `py` is necessary since `Py` can exist even after we release GIL.

Reply:
This sounds reasonable but actually now I search the codebase, in both of the implementations of `AsPyRef` we don't actually use the `py` argument. I propose we remove the trait completely, as it's essentially equivalent to `AsRef`.

The two implementations I see are:

https://github.com/kngwyu/pyo3/blob/pyclass-new-layout/src/instance.rs#L127

https://github.com/kngwyu/pyo3/blob/pyclass-new-layout/src/object.rs#L253